### ISSUE TITLE
`getAll` with function-based singletons only returns a single instance

### ISSUE THREAD

**Issue by SIMULATAN** [InsertKoinIO/koin-annotations#165]
**Describe the bug**
Using `Koin#getAll()` to get `@Single` annotated functions sharing a return type only returns a single instance. The others get lost during registration and don't even show up in the internal Koin `_instances` map.

**To Reproduce**
1. Create a service interface
2. Implement it anonymously in an arbitrary (2+) number of function bodies
3. Annotate those functions with `@Single`
4. Attempt to use `getAll` - only one function is returned

Please find a dead simple reproducer here: https://github.com/SIMULATAN/koin-getall-reproducer

**Expected behavior**
All instances are returned.

**Koin project used and used version**
koin-core 4.0.0-RC1
koin-annotations 1.4.0-RC4

**Analysis**
From my analysis, the problem is that koin-annotations registers the singletons using the same `binds` class: the function return type. This poses an issue with koin's internal `instanceRegistry._instances` map, which stores the instances using the class name as the key. As established, this class name is the interface type, which causes an arbitrary number of singletons with that type to only result in a single getable instance.
![reproducer in intellij with the debugger showing the internal koin map](https://github.com/user-attachments/assets/a6eda735-c4be-41c7-b39a-534ea2f22725)
In short: using subclasses, `getAll` works as koin saves the implementations by their concrete class name in the internal map. However, with `@Single fun`, the "concrete class" becomes the superclass, resulting in overrides in the internal map.

**Proposed solution**
Naming beans created from functions, for example based on the function name. Although not perfect, this may prevent collisions and it serves a similar system as class implementations.

**Comment by arnaudgiuliani** [comment]
Interesting. Can you just share the generated DSL? A definition should allow several type binding. Annotations should take original class and second type binding

**Comment by SIMULATAN** [comment]
```kt
@Definition("me.simulatan.koin.getall")
public fun Module.defineFirstResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.FirstResource() } bind(me.simulatan.koin.getall.Resource::class)
@Definition("me.simulatan.koin.getall")
public fun Module.defineSecondResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.SecondResource() } bind(me.simulatan.koin.getall.Resource::class)
@Definition("me.simulatan.koin.getall")
public fun Module.definefirstResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.firstResource() } bind(me.simulatan.koin.getall.Resource::class)
@Definition("me.simulatan.koin.getall")
public fun Module.definesecondResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.secondResource() } bind(me.simulatan.koin.getall.Resource::class)
@Definition("me.simulatan.koin.getall")
public fun Module.definethirdResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.thirdResource() } bind(me.simulatan.koin.getall.Resource::class)
```
(I removed the boilerplate)

**Comment by SIMULATAN** [comment]
Just ran into this issue again. Any chance to get this fixed soon-ish?
I'm absolutely willing to contribute a patch, would auto-naming function-based definitions using the function's name suffice or would you suggest another approach?

**Comment by arnaudgiuliani** [comment]
Sorry for delay. Need to reevaluate in KA 2.0 branch. Any chance for you to update your sample to Koin 4 + Koin Annotations 2? @SIMULATAN

**Comment by SIMULATAN** [comment]
@arnaudgiuliani sure! https://github.com/SIMULATAN/koin-getall-reproducer/commit/374ba39b2eef1e21f5d6ba9047c4ea4b0f9b0305

**Comment by arnaudgiuliani** [comment]
Need to double check if it's not related to Koin core internals issue

**Comment by arnaudgiuliani** [comment]
> Just ran into this issue again. Any chance to get this fixed soon-ish? I'm absolutely willing to contribute a patch, would auto-naming function-based definitions using the function's name suffice or would you suggest another approach?

@SIMULATAN 
Just a question: Do those functions are used in your main module? Those generated extensions, as exported definitions, should be used in a more central module.

I need to check your sample project.

**Comment by SIMULATAN** [comment]
Yep, they are in my `defaultModule`. Currently, I'm using this workaround.
```kt
@Single
@Named("about")
fun AboutResource(buildInfo: BuildInfo) = Resource {
	// ...
}
```
This results in these extensions:
```kt
@Definition("me.snoty.backend.server.resources")
public fun Module.defineAboutResource() : KoinDefinition<*> = single(qualifier=org.koin.core.qualifier.StringQualifier("about")) { _ -> me.snoty.backend.server.resources.AboutResource(buildInfo=get()) } bind(me.snoty.backend.server.routing.Resource::class)
@Definition("me.snoty.backend.server.resources")
public fun Module.definewiringResources() : KoinDefinition<*> = single(qualifier=org.koin.core.qualifier.StringQualifier("wiring")) { _ -> me.snoty.backend.server.resources.wiringResources(hookRegistry=get()) } bind(me.snoty.backend.server.routing.Resource::class)
@Definition("me.snoty.backend.server.resources.task")
public fun Module.definetaskResources() : KoinDefinition<*> = single(qualifier=org.koin.core.qualifier.StringQualifier("task")) { _ -> me.snoty.backend.server.resources.task.taskResources(adminTasks=get()) } bind(me.snoty.backend.server.routing.Resource::class)
```

These resources are defined here: https://github.com/SnotyMe/snoty-backend/tree/65931b36fe9184944ffddd59818abc787bd2089e/src/main/kotlin/me/snoty/backend/server/resources

**Comment by arnaudgiuliani** [comment]
> @Definition("me.simulatan.koin.getall")
> public fun Module.defineFirstResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.FirstResource() } bind(me.simulatan.koin.getall.Resource::class)
> @Definition("me.simulatan.koin.getall")
> public fun Module.defineSecondResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.SecondResource() } bind(me.simulatan.koin.getall.Resource::class)
> @Definition("me.simulatan.koin.getall")
> public fun Module.definefirstResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.firstResource() } bind(me.simulatan.koin.getall.Resource::class)
> @Definition("me.simulatan.koin.getall")
> public fun Module.definesecondResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.secondResource() } bind(me.simulatan.koin.getall.Resource::class)
> @Definition("me.simulatan.koin.getall")
> public fun Module.definethirdResource() : KoinDefinition<*> = single() { me.simulatan.koin.getall.thirdResource() } bind(me.simulatan.koin.getall.Resource::class)
> (I removed the boilerplate)

I see the problem.  You are shadowing the same definition at the end. from your sample project, if I update names from classes I can get more stuff: 

```kotlin
@Single
class FirstClassResource : Resource {
	override fun invoke() {
		println("First class resource")
	}
}

@Single
class SecondClassResource : Resource {
	override fun invoke() {
		println("Second class resource")
	}
}
```

in the end, you need qualifier on your resource function to help make different implementations 👍

**Comment by arnaudgiuliani** [comment]
I close the issue but we can continue to discuss if needed

**Comment by SIMULATAN** [comment]
> You are shadowing the same definition at the end. from your sample project, if I update names from classes I can get more stuff

Yep, that was the reason why I created this issue. `@Single` annotated *classes* are qualified by the concrete subclass name. `@Single` annotated *functions* however are only registered under the *return type name*, thus overwriting one another.
I'd propose to change the key under which the dependency is registered in `InstanceRegistry#_instances` to `package.function` to align the behavior with classes and prevent overwrites. IMO, that would certainly work as a unique qualifier as public function names are unique in the scope of a Kotlin package anyway. Bindings for the return type should be unaffected.

**Comment by arnaudgiuliani** [comment]
@Terenfear it's tied to annotations too?

**Comment by Terenfear** [comment]
@arnaudgiuliani not sure I understand your question, sorry. 

I use annotated class modules with annotated function providers (e.g. `@Single fun provideFoo() = Foo()`). I have quite a lot of them and some, by design, provide the same type while having a unique function name. 

Then I have dependencies that try to get all instances of some type via a list-typed parameter (which under the hood uses `getAll`). 
I expected it to work kinda like it does in Dagger with `@IntoMap` multibinding, but without the extra step of providing a class key. AFAIU it does work like that in Koin, but only with `@Single` annotated classes, as @SIMULATAN mentioned.

I was surprised that it doesn't work like that with annotated functions, because I think there's enough info to make a unique key for each binding. Currently I use `@Named` workaround proposed by @SIMULATAN, but I'd certainly prefer if Koin were able to figure out that multiple functions with unique names providing the same type should not override each other.
